File: n_queens.py
from jutge import read
'''
N-queens
'''

class Board:
    def __init__(self, N):
        self.N = N
        self.Chess = dict()

        for row in range(N):
            self.Chess[row] = N*[0]

    def __str__(self):
        for k in self.Chess:
            print(str(self.Chess[k]))

    def free(self, row, col):
        for i in range(row):
            if self.Chess[i][col]:
                return False
        for j in range(col):
            if self.Chess[row][j]:
                return False
        for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
            if self.Chess[i][j]: return False
        for i, j in zip(range(row, -1, -1), range(col, self.N)):
            if self.Chess[i][j]: return False
        return True

    def put_q(self, row, col):
        self.Chess[row][col] = 1

    def remove_Q(self, row, col):
        self.Chess[row][col] = 0

def attempt(row, board, size):
    if row == size:
        board.draw() #This function has not been implemented yet.
    else:
        for column in range(size):
            if board.free(row, column):
                board.put_q(row, column) #put the queen
                attempt(row+1, board, size) #skipping the current row, to move down in the tree
                board.remove_Q(row, column) #remove the queen because we are doing the for loop by column, so it will keep tracking all possibilities

def attempt2(row, board, size):
    if row == size: #same queens than rows
        return True ##if I get to the end return true
    else:
        for column in range(size): #check all columns
            if board.free(row, column):
                board.put_q(row, column)
                s = attempt2(row + 1, board, size) #value that returns true (row = size)
                if s:
                    return True
                else:
                    board.remove_Q(row, column)
        return False #non of the branches found a solution


print('How many queens? ')
N = read(int)
board = Board(N)
print("The empty chess board: \n")
board.__str__()
print()
attempt2(0, board, N)
print()
print("One solution of the N-Queen: \n ")
board.__str__()

File: n_queens.cc
// snippet of a class to implement the n times n chess board for the n-queens problem
#include<iostream>
#include<vector>
using namespace std;


class Board {
private:
   // insert the code of the representation here
   // representation of the board, MATRIX, booleans
   int n;
   vector<vector<int> > b;

public:

   Board(int n): b(n, vector<int>(n))
   {
       //b = vector<vector<int> > (n, vector<int>(n));
       for (int i =0; i < n; ++i){
           for(int j = 0; j < n; ++j){
               b[i][j] = 0;
           }
       }
       print_c(n);
   }

   //void draw() const {

   //}
   void print_c(int n)
   {
       cout << "ChessBoard representation" << endl;
       for (int x =0; x < n; ++x){
          for(int j = 0; j < n; ++j){
              cout << b[x][j] << " ";
          }
          cout << endl;
       }
   }
   bool free(int row, int column) {
       for(int i=0; i < row; ++i){
           if (b[i][column]) return false;
       }
       for(int j=0; j < column; ++j){
           if(b[row][j]) return false;
       }
       for (int i=row, j=column; i>=0 && j>=0; i--, j--)
          if (b[i][j]) //check whether there is queen in the left upper diagonal or not
             return false;
       for (int i=row, j=column; j>=0 && i<b.size(); i--, j++)
          if (b[i][j]) //check whether there is queen in the left lower diagonal or not
             return false;
       // check the diagonals now
       return true;
   }

   void put_q(int row, int column){
       b[row][column] = 1;
   }

   void remove_q(int row, int column){
       b[row][column] = 0;
   }
};

bool attempt(int row, Board& board, int size){
    int i;
    if (row == size) return true;
    else{
        for (i = 0; i < size; ++i){
            if (board.free(row, i)){
                board.put_q(row, i);
                bool s = attempt(row+1, board, size);
                if(s) return true;
                else board.remove_q(row, i);
            }
        }
        return false;
    }
}

int main(){
    int N;
    cout << "Enter the number of Queens: ";
    cin >> N;
    Board ChessBoard(N);
    attempt(0, ChessBoard, N);
    ChessBoard.print_c(N);
;
}
